### PR TITLE
RavenDB-8924 Adding support for RegexQuery

### PR DESCRIPTION
**Pull request by talweiss1982**

### CODE REVIEW

File: src/Raven.Server/Documents/Queries/QueryBuilder.cs [hunk 1]
```diff
@@ -424,6 +427,23 @@ private static Lucene.Net.Search.Query HandleEndsWith(Query query, MethodExpress
...
+            BlittableJsonReaderObject parameters)
+        {
+
+            var regex = ((ValueExpression)expression.Arguments[1]).Token.Value;
+
+            var fieldName = ExtractIndexFieldName(query, parameters, expression.Arguments[0], metadata);
+            var (value, valueType) = GetValue(fieldName, query, metadata, parameters, (ValueExpression)expression.Arguments[1]);
+
+            if (valueType != ValueTokenType.String)
+                ThrowMethodExpectsArgumentOfTheFollowingType("regex", ValueTokenType.String, valueType, metadata.QueryText, parameters);
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/QueryBuilder.cs:440]: Need to also support parameter here.

File: src/Raven.Server/Documents/Queries/QueryBuilder.cs [hunk 2]
```diff
@@ -424,6 +427,23 @@ private static Lucene.Net.Search.Query HandleEndsWith(Query query, MethodExpress
             return q;
         }
 
+        private static Lucene.Net.Search.Query HandleRegex(Query query, MethodExpression expression, QueryMetadata metadata,
+            BlittableJsonReaderObject parameters)
+        {
+
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/QueryBuilder.cs:433]: Validate number of arguments.

File: src/Raven.Server/Documents/Queries/Regex/RegexQuery.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    /// </summary>
+    /// <remarks>http://www.java2s.com/Open-Source/Java-Document/Net/lucene-connector/org/apache/lucene/search/regex/RegexQuery.java.htm</remarks>
+    public class RegexQuery : MultiTermQuery, IEquatable<RegexQuery>
+    {
+        public Term Term { get; }
+
+        //Increasing .net regex cache size if not modified by another code path
+        static RegexQuery()
+        {
+            Regex.CacheSize = Regex.CacheSize == 15 ? 50 : Regex.CacheSize;
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/Regex/RegexQuery.cs:38]: Kill it.

File: src/Raven.Server/Documents/Queries/Regex/RegexTermEnum.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+            SetEnum(reader.Terms(new Term(term.Field, string.Empty), state), state);
+        }
+
+        /// <summary>Equality compare on the term </summary>
+        protected override bool TermCompare(Term term)
+        {
+            if (_sField == term.Field)
+            {
+                return Regex.IsMatch(term.Text, _pattern);
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/Regex/RegexTermEnum.cs:52]: Have a `Regex` instance that we hold on to.
The logic goes like this.
A concurrent dictionary of up to 1024 items (configuration).
If exceeding that, evict random item from cache.
If cache is full, and number of regex creations per minute exceed (configuration), clear the cache, raise alert, and then move to using non compiled regex.

File: test/FastTests/Client/Queries/RegexQueryTest.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+                using (var session = store.OpenSession())
+                {
+                    session.Store(new RegexMe{Text = "I love dogs and cats"});
+                    session.Store(new RegexMe { Text = "I love cats" });
+                    session.Store(new RegexMe { Text = "I love dogs" });
+                    session.Store(new RegexMe { Text = "I love bats" });
+                    session.Store(new RegexMe { Text = "dogs love me" });
+                    session.Store(new RegexMe { Text = "cats love me" });
+                    session.SaveChanges();
+                    var res = session.Advanced.RawQuery<RegexMe>("from RegexMes as r where Regex(r.Text,\"^[a-z ]{2,4}love\")")
```
Comment (ayende) [inline test/FastTests/Client/Queries/RegexQueryTest.cs:25]: Create issue to support this from the Linq provider

File: src/Raven.Server/Documents/Queries/Regex/RegexTermEnum.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/Documents/Queries/Regex/RegexTermEnum.cs:52]: Let's set eviction policy on least recently used and handle too many at the same time.

File: test/FastTests/Client/Queries/RegexQueryTest.cs (hunk 5 above)
Comment (talweiss1982) [reply test/FastTests/Client/Queries/RegexQueryTest.cs:25]: http://issues.hibernatingrhinos.com/issue/RavenDB-9094

**Pull request merged**